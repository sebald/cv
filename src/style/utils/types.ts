/**
 * The values will be treated like the following:
 * - number -> "px" value
 * - string -> "rem" value
 */
export type FontValue = number | string;

export type ScaleConfig = {
  fontSize: FontValue;
  htmlFontSize: number;
  ratio: number;
};

export type LineHeightConfig = {
  fontSize: FontValue;
  htmlFontSize: number;
  baseline: number;
};

export type TypographyConfig = ScaleConfig & LineHeightConfig;

export type WithClassName = {
  className?: string;
};
