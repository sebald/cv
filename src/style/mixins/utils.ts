import { CSSProperties } from 'react';
import { StyledMixin } from '../theme';

export const propOrEmpty = <P extends keyof CSSProperties>(
  name: P
): StyledMixin<Pick<CSSProperties, P>> => (props: Pick<CSSProperties, P>) => {
  const p = props[name];
  return p ? { [name]: p } : {};
};
