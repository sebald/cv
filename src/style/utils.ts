import { theme } from './theme';

const { htmlFontSize } = theme;

export const px = (value: number) => `${value}px`;

export const pxToRem = (pixel: number, base = htmlFontSize) =>
  `${parseFloat((pixel / base).toFixed(5))}rem`;

export const remToPx = (rem: string, base = htmlFontSize) =>
  parseFloat(rem) * base;

export const ensurePx = (value: number | string, base = htmlFontSize) =>
  typeof value === 'string' ? remToPx(value, base) : value;
