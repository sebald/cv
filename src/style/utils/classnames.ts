export const mergeClassNames = (...names: (string | undefined)[]) => ({
  className: names
    .filter(x => typeof x === 'string')
    .join(' ')
    .trim(),
});
