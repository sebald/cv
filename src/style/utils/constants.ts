export const htmlFontSize = 16;

export const ratio = {
  perfectFourth: 1.333,
  perfectFifth: 1.5,
  goldenRatio: 1.618,
  majorSixth: 1.667,
};
