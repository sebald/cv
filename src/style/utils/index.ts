export * from './classnames';
export * from './constants';
export * from './createMixinComponent';
export * from './interpolation';
export * from './unit';
export * from './typography';
