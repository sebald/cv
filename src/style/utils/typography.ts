import { LineHeightConfig, ScaleConfig, TypographyConfig } from './types';
import { ensurePx, pxToRem } from './unit';

/**
 * Create a function that generates modular scales base on given `fontSize`,
 * `htmlFontSize` and `ratio`. The resuling scale (font size value) will be
 * in `rem`.
 */
export const createScale = ({ fontSize, htmlFontSize, ratio }: ScaleConfig) => (
  level: number
) => pxToRem(ensurePx(fontSize, htmlFontSize) * ratio ** level, htmlFontSize);

/**
 * Returns the next line height (in "rem") for a given `fontSize` that will satisfy
 * vertical rhythm.
 *
 * Usually, in CSS you'll use a unitless number for the value of `line-height`
 * in order to avoid unwanted line-height cascading for nested elements containing text.
 * However, in this system, all elements containing text should/will have an explicit line
 * height set, so they won't be influenced by the cascade.
 *
 * For more information see: http://meyerweb.com/eric/thoughts/2006/02/08/unitless-line-heights/
 */
export const getLineHeight = ({
  fontSize,
  htmlFontSize,
  baseline,
}: LineHeightConfig) =>
  pxToRem(
    Math.ceil(
      ensurePx(ensurePx(fontSize, htmlFontSize), htmlFontSize) / baseline
    ) * baseline,
    htmlFontSize
  );

/**
 * Returns a `function` that generates typographic values (`fontSize`, `lineHeight`).
 * The generated values adhere configured (typograpic) `scale` and `baseline`.
 */
export const createTypography = ({
  fontSize,
  htmlFontSize,
  ratio,
  baseline,
}: TypographyConfig) => {
  const scale = createScale({ fontSize, htmlFontSize, ratio });
  return (level: number) => {
    const fs = scale(level);
    return {
      fontSize: fs,
      lineHeight: getLineHeight({ fontSize: fs, htmlFontSize, baseline }),
    };
  };
};

/**
 * Small helper function to align text with the base line.
 *
 * Why are we doing this? CSS's `line-height` will sadly not align text to the
 * typograpic baseline, but rather will vertically center the font. So if we are
 * really pendantic (which we are BTW ... at least this time), we need to
 *
 * (1) add some padding at the top (to move the text to the baseline) and
 * (2) add some margin to the bottom to preserver verticyl rhythm
 *
 * You can read more information about this approach here:
 * https://www.smashingmagazine.com/2012/12/css-baseline-the-good-the-bad-and-the-ugly/
 */
export const adjustBaseline = (baseline: number) => (
  top?: number,
  factor = 1
) => ({
  ...(top && { paddingTop: pxToRem(top) }),
  marginTop: 0,
  marginBottom: pxToRem(factor * baseline - (top || 0)),
});
