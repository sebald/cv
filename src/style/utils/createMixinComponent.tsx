import React from 'react';

import { styled } from '../styled-components';
import { StyledMixin } from '../theme';

import { mergeClassNames } from './classnames';
import { WithClassName } from './types';

/**
 * Idea stolen from:
 * https://github.com/jxnblk/grid-styled/tree/master/styled-space
 */
const StyledChildren: React.FC<WithClassName> = ({ className, children }) => {
  const styledChildren = React.Children.toArray(children).map((child) =>
    React.isValidElement<any>(child)
      ? React.cloneElement(
          child,
          mergeClassNames(child.props.className, className)
        )
      : child
  );

  return <>{styledChildren}</>;
};

export const createMixinComponent = <P extends {}>(
  ...mixins: StyledMixin<P>[]
) => styled(StyledChildren)([] as any, ...mixins);
