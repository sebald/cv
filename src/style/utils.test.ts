import { ensurePx, px, pxToRem, remToPx } from './utils';

test('px', () => {
  expect(px(16)).toMatchSnapshot();
  expect(px(1)).toMatchSnapshot();
  expect(px(24)).toMatchSnapshot();
  expect(px(0)).toMatchSnapshot();
});

test('px to rem (using base font size)', () => {
  expect(pxToRem(16)).toMatchSnapshot();
  expect(pxToRem(1)).toMatchSnapshot();
  expect(pxToRem(20)).toMatchSnapshot();
  expect(pxToRem(0)).toMatchSnapshot();
});

test('px to rem (custom base font size)', () => {
  expect(pxToRem(16, 10)).toMatchSnapshot();
  expect(pxToRem(1, 10)).toMatchSnapshot();
  expect(pxToRem(20, 20)).toMatchSnapshot();
});

test('rem to px (using base font size)', () => {
  expect(remToPx('1rem')).toMatchSnapshot();
  expect(remToPx('1.25rem')).toMatchSnapshot();
  expect(remToPx('0.75rem')).toMatchSnapshot();
  expect(remToPx('0rem')).toMatchSnapshot();
});

test('rem to px (custom base font size)', () => {
  expect(remToPx('1rem', 12)).toMatchSnapshot();
  expect(remToPx('1.25rem', 10)).toMatchSnapshot();
  expect(remToPx('0.75rem', 16)).toMatchSnapshot();
  expect(remToPx('0rem', 10)).toMatchSnapshot();
});

test('ensure value is px (using base font size)', () => {
  expect(ensurePx(16)).toMatchSnapshot();
  expect(ensurePx('1rem')).toMatchSnapshot();
  expect(ensurePx(24)).toMatchSnapshot();
  expect(ensurePx('1.25rem')).toMatchSnapshot();
});

test('ensure value is px (custom base font size)', () => {
  expect(ensurePx(16, 10)).toMatchSnapshot();
  expect(ensurePx('1rem', 20)).toMatchSnapshot();
  expect(ensurePx(20, 12)).toMatchSnapshot();
  expect(ensurePx('1.25rem', 30)).toMatchSnapshot();
});
