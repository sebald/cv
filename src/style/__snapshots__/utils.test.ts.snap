// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensure value is px (custom base font size) 1`] = `16`;

exports[`ensure value is px (custom base font size) 2`] = `20`;

exports[`ensure value is px (custom base font size) 3`] = `20`;

exports[`ensure value is px (custom base font size) 4`] = `37.5`;

exports[`ensure value is px (using base font size) 1`] = `16`;

exports[`ensure value is px (using base font size) 2`] = `16`;

exports[`ensure value is px (using base font size) 3`] = `24`;

exports[`ensure value is px (using base font size) 4`] = `20`;

exports[`px 1`] = `"16px"`;

exports[`px 2`] = `"1px"`;

exports[`px 3`] = `"24px"`;

exports[`px 4`] = `"0px"`;

exports[`px to rem (custom base font size) 1`] = `"1.6rem"`;

exports[`px to rem (custom base font size) 2`] = `"0.1rem"`;

exports[`px to rem (custom base font size) 3`] = `"1rem"`;

exports[`px to rem (using base font size) 1`] = `"1rem"`;

exports[`px to rem (using base font size) 2`] = `"0.0625rem"`;

exports[`px to rem (using base font size) 3`] = `"1.25rem"`;

exports[`px to rem (using base font size) 4`] = `"0rem"`;

exports[`rem to px (custom base font size) 1`] = `12`;

exports[`rem to px (custom base font size) 2`] = `12.5`;

exports[`rem to px (custom base font size) 3`] = `12`;

exports[`rem to px (custom base font size) 4`] = `0`;

exports[`rem to px (using base font size) 1`] = `16`;

exports[`rem to px (using base font size) 2`] = `20`;

exports[`rem to px (using base font size) 3`] = `12`;

exports[`rem to px (using base font size) 4`] = `0`;
